### ISSUE TITLE
Enforcing Partition limits

### ISSUE THREAD

**Issue by SamBarker** [kroxylicious/design#9]
## Motivation

Enforcing a partition limits across a Kafka cluster is a desirable goal to allow cluster operators to ensure a quality of service to their users. However given the current Kafka API's for managing partitions within a single broker limits are enforced in a best efforts fashion with several corner cases which allow the limits to be breached. 

We see partition limit enforcement as emblematic for a class of use cases for a Kafka aware proxy. Focusing on a specific use cases allows us to design and build a functional system which can be deployed to gain a deeper understanding quickly.

## Use cases
1. Enforce limits at topic creation.
2. Enforce limits when changing topic configuration.

## Assuming:
- A cluster of proxy instances. `p1`, `p2`, `p3` 
- A cluster of Kafka Brokers. `b1`, `b2`, `b3` 
- A Kafka client (`c1`) connected & authenticated with `b3` via a proxy instance (`p2`)
- The cluster operator has defined a limit to the number of Kafka partitions.

**Comment by showuon** [comment]
Nice proposal! 
I think the cons of leader solution has one more, which is when leader proxy crashed, it needs time to elect another leader, and sync up with kafka clusters for `ActivePartitions`, before it can start working. Is my understanding correct?

**Comment by tombentley** [comment]
@showuon that's correct. It's worth noting, however, that it should be less of a problem than Kafka controller elections, because it's only client-broker topic-related requests; things like ISR change etc needn't be affected.
